File: Renderer/src/main/java/darwin/renderer/geometrie/attributs/VAOAttributs.java
/*
 * Copyright (C) 2012 daniel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package darwin.renderer.geometrie.attributs;

import darwin.renderer.GraphicContext;
import darwin.renderer.opengl.VertexBO;
import darwin.renderer.opengl.buffer.BufferObject;
import darwin.renderer.shader.Shader;

import com.google.inject.assistedinject.*;
import javax.media.opengl.GLProfile;

/**
 *
 ** @author Daniel Heinrich <DannyNullZwo@gmail.com>
 */
public final class VAOAttributs implements AttributsConfigurator {

    public interface VAOAttributsFactory {

        public VAOAttributs create(Shader shader, VertexBO[] vbuffers,
                                   BufferObject indice);
    }

    static {
        assert GLProfile.isAvailable(GLProfile.GL2GL3) : "This device VertexAttributsdoesn't support VAOs";
    }
    private final GraphicContext gc;
    private final int id;

    @AssistedInject
    public VAOAttributs(GraphicContext gcontext,
                        @Assisted Shader shader,
                        @Assisted VertexBO[] vbuffers,
                        @Assisted  BufferObject indice) {
        gc = gcontext;

        int[] i = new int[1];
        gc.getGL().getGL2GL3().glGenVertexArrays(1, i, 0);
        id = i[0];

        StdAttributs sa = new StdAttributs(gc, shader, vbuffers, indice);
        prepare();
        sa.prepare();
        disable();
        //TODO: wtf??? GL Error about vbo not bound
        sa.disable();
    }

    @Override
    public void prepare() {
        gc.getGL().getGL2GL3().glBindVertexArray(id);
    }

    @Override
    public void disable() {
        gc.getGL().getGL2GL3().glBindVertexArray(0);
    }
}

File: Renderer/src/main/java/darwin/renderer/geometrie/attributs/VAOAttributs.scala
/*
 * Copyright (C) 2012 daniel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package darwin.renderer.geometrie.attributs

import darwin.renderer.{GProfile, Bindable}
import darwin.renderer.opengl.GLResource
import darwin.renderer.shader.ShaderComponent
import javax.media.opengl.GL2GL3
import darwin.renderer.dependencies.Components

/**
 *
 * * @author Daniel Heinrich <DannyNullZwo@gmail.com>
 */
trait VAOComponent extends StdAttributesComponent {
  this: Components#VBO with GProfile[GL2GL3] with ShaderComponent=>

  import context._

  override def createAttributes(shader: Shader, vbuffers: Seq[VertexBO], indice: Option[BufferObject]) = {
    val i: Array[Int] = new Array[Int](1)
    gl.glGenVertexArrays(1, i, 0)

    val vao = new VAOAttributs(i(0))
    val sa = super.createAttributes(shader, vbuffers, indice)

    vao.bind
    sa.bind
    vao.unbind
    sa.unbind

    vao
  }


  class VAOAttributs(val id: Int) extends GLResource with Bindable {
    def bind {
      gl.glBindVertexArray(id)
    }

    def unbind {
      gl.glBindVertexArray(0)
    }

    def delete() {
      gl.glDeleteVertexArrays(1, Array(id), 0)
    }
  }

}